File: apps/mobile/shared/Sources/FlagshipAPI/Client/PhoneAlertClient.swift
import Foundation

/// #91 — drains the daemon→phone alert outbox over the paired-session pinned
/// pipe (`GET /api/phone/alerts` + `POST /api/phone/alerts/ack`). Same auth as
/// the screens client (the 32-byte hex session token in `x-flagship-session`);
/// the box terminates TLS, so flagshipserver.com never sees these.
public protocol PhoneAlertClient: Sendable {
    /// Fetch alerts queued after `since` (the last-seen id; 0 = from the start).
    func fetchAlerts(since: Int) async throws -> PhoneAlertsResponse
    /// Acknowledge / drop alerts through `throughId` so they aren't re-delivered.
    func ackAlerts(throughId: Int) async throws
}

/// URLSession-backed `PhoneAlertClient`, mirroring `LiveScreensClient`'s
/// request shape (and reusing its `SessionStoring` + the box-pinned session).
public final class LivePhoneAlertClient: PhoneAlertClient, @unchecked Sendable {
    private let urlSession: URLSession
    private let store: any SessionStoring

    public init(urlSession: URLSession = .shared, store: any SessionStoring) {
        self.urlSession = urlSession
        self.store = store
    }

    public func fetchAlerts(since: Int) async throws -> PhoneAlertsResponse {
        try await request("/api/phone/alerts?since=\(since)")
    }

    public func ackAlerts(throughId: Int) async throws {
        let body = try JSONSerialization.data(withJSONObject: ["throughId": throughId])
        let _: AckResponse = try await request("/api/phone/alerts/ack", method: "POST", body: body)
    }

    private struct AckResponse: Decodable { let ok: Bool? }

    private func request<T: Decodable>(_ path: String, method: String = "GET", body: Data? = nil) async throws -> T {
        guard let base = await store.podBaseUrl else { throw ScreensClientError.notPaired }
        guard let token = await store.sessionToken else { throw ScreensClientError.noSessionToken }
        guard let url = URL(string: base.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + path) else {
            throw ScreensClientError.http(status: 0, message: "bad URL")
        }

        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue(token, forHTTPHeaderField: "x-flagship-session")
        if let body {
            req.httpBody = body
            req.setValue("application/json", forHTTPHeaderField: "content-type")
        }

        let (data, resp) = try await urlSession.data(for: req)
        guard let http = resp as? HTTPURLResponse else {
            throw ScreensClientError.http(status: 0, message: "no response")
        }
        if !(200..<300).contains(http.statusCode) {
            throw ScreensClientError.http(status: http.statusCode, message: String(data: data, encoding: .utf8) ?? "")
        }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            throw ScreensClientError.decoding(String(describing: error))
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/api/PhoneAlertClient.kt
// #91 — drains the daemon→phone alert outbox over the paired-session pipe
// (`GET /api/phone/alerts` + `POST /api/phone/alerts/ack`). Same auth as the
// screens client (the 32-byte hex session token in `x-flagship-session`); the
// box terminates TLS, so flagshipserver.com never sees these.
//
// Kotlin mirror of apps/mobile/shared/.../FlagshipAPI/Client/PhoneAlertClient.swift.

package com.flagshipserver.app.api

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext

interface PhoneAlertClient {
    /** Fetch alerts queued after [since] (the last-seen id; 0 = from the start). */
    suspend fun fetchAlerts(since: Int): PhoneAlertsResponse
    /** Acknowledge / drop alerts through [throughId] so they aren't re-delivered. */
    suspend fun ackAlerts(throughId: Int)
}

/** OkHttp-backed [PhoneAlertClient], mirroring [LiveScreensClient]'s request
 *  shape (and reusing its [SessionStoring] + the same OkHttp client, which the
 *  app builds with the box cert-pinning interceptor). */
class LivePhoneAlertClient(
    private val client: OkHttpClient = OkHttpClient(),
    private val store: SessionStoring,
) : PhoneAlertClient {

    private val jsonMt = "application/json; charset=utf-8".toMediaType()

    private suspend fun base(): String = store.podBaseUrl.first()?.trimEnd('/')
        ?: throw ScreensError.NotPaired

    private suspend fun token(): String = store.sessionToken.first()
        ?: throw ScreensError.NoSessionToken

    override suspend fun fetchAlerts(since: Int): PhoneAlertsResponse {
        val body = get("/api/phone/alerts?since=$since")
        return PhoneAlertsResponse.parse(body)
    }

    override suspend fun ackAlerts(throughId: Int) {
        post("/api/phone/alerts/ack", "{\"throughId\":$throughId}")
    }

    private suspend fun get(path: String): String {
        val req = Request.Builder()
            .url(base() + path)
            .header("x-flagship-session", token())
            .get()
            .build()
        return execute(req)
    }

    private suspend fun post(path: String, jsonBody: String): String {
        val req = Request.Builder()
            .url(base() + path)
            .header("x-flagship-session", token())
            .header("content-type", jsonMt.toString())
            .post(jsonBody.toRequestBody(jsonMt))
            .build()
        return execute(req)
    }

    private suspend fun execute(req: Request): String {
        val (status, bytes) = withContext(Dispatchers.IO) {
            client.newCall(req).execute().use { resp ->
                resp.code to (resp.body?.bytes() ?: ByteArray(0))
            }
        }
        if (status !in 200..299) throw ScreensError.Http(status, String(bytes, Charsets.UTF_8))
        return String(bytes, Charsets.UTF_8)
    }
}
